projects/Mallard/src/services/logging.ts

```diff
@@ -0,0 +1,190 @@
+            iapReceipt,
+        ] = await Promise.all([
+            NetInfo.fetch(),
+            userDataCache.get(),
+            getCASCode(),
+            iapReceiptCache.get(),
+        ])
+        return {
+            networkStatus,
+            userData,
+            casCode,
+            iapReceipt,
+        }
+    }
+
+    async baseLog({
+        level,
+        message,
+        ...optionalFields
+    }: LogParams): Promise<MallardLogFormat> {
+        const {
+            networkStatus,
+            userData,
+            casCode,
+            iapReceipt,
+        } = await this.getExternalInfo()
+
+        // User Data and Subscription
+        const userId =
+            (userData && userData.userDetails && userData.userDetails.id) || ''
+        const digitalSub =
+            (userData &&
+                userData.membershipData &&
+                userData.membershipData.contentAccess &&
+                userData.membershipData.contentAccess.digitalPack) ||
+            false
+        const iAP = iapReceipt ? true : false
+
+        return {
+            app: DeviceInfo.getBundleId(),
+            version: DeviceInfo.getVersion(),
+            buildNumber: DeviceInfo.getBuildNumber(),
+            os: Platform.OS === 'ios' ? 'ios' : 'android',
+            device: DeviceInfo.getDeviceId(),
+            networkStatus: networkStatus
+                ? networkStatus.type
+                : NetInfoStateType.unknown,
+            release_channel: isInBeta() ? 'BETA' : 'RELEASE',
+            timestamp: new Date(),
+            level,
+            message,
+            deviceId: DeviceInfo.getUniqueId(),
+            signedIn: userData ? true : false,
+            userId,
+            digitalSub,
+            casCode,
+            iAP,
+            ...optionalFields,
+        }
+    }
+
+    async getQueuedLogs(): Promise<MallardLogFormat[] | [{}]> {
+        try {
+            const logString = await loggingQueueCache.get()
+            return JSON.parse(logString || '[{}]')
+        } catch (e) {
+            return [{}]
+        }
+    }
+
+    async saveQueuedLogs(log: MallardLogFormat[]): Promise<string | Error> {
+        try {
+            const logString = JSON.stringify(log)
+            await loggingQueueCache.set(logString)
+            return 'saved logs'
+        } catch (e) {
+            errorService.captureException(e)
+            throw new Error(e)
+        }
+    }
+
+    async queueLogs(log: MallardLogFormat[]) {
+        try {
+            const currentQueue = await this.getQueuedLogs()
+            const currentQueueString = JSON.stringify(currentQueue)
+            const parsedQueue = JSON.parse(currentQueueString || '[{}]')
+            const newQueue = [...parsedQueue, ...log]
+            const cleanLogs = newQueue.filter(
+                value => Object.keys(value).length !== 0,
+            )
+            return cleanLogs
+        } catch (e) {
+            errorService.captureException(e)
```

Comment:
If it fails once that would be due to malformed data structure and it will probably keep failing. Shall we consider clearing the logs if it fails (maybe after X attempts or something like that)?

Reply:
Definitely sounds like a good optimisation. Will add a ticket to cover these scenarios